src/main/java/org/candlepin/pki/impl/BouncyCastleKeyPairGenerator.java

```diff
@@ -0,0 +1,254 @@
+            keyGen.initialize(KEY_SIZE);
+            return keyGen.generateKeyPair();
+        }
+        catch (NoSuchAlgorithmException e) {
+            throw new KeyPairCreationException("Failed to generate a new key pair!", e);
+        }
+    }
+
+    /**
+     * Attempts to process the given keypair data as if the keys are PKCS8 formatted. If the key
+     * pair data is incomplete or invalid, this method returns null.
+     *
+     * @param kpdata
+     *  the key pair data to process
+     *
+     * @return
+     *  a KeyPair consisting of the keys from the provided key pair data, or null if the key pair
+     *  data could not be processed
+     */
+    private KeyPair processAsPKCS8(KeyPairData kpdata) {
+        try {
+            PublicKey publicKey = this.generatePublicKey(kpdata.getPublicKeyData(), KEY_ALGORITHM);
+            PrivateKey privateKey = this.generatePrivateKey(kpdata.getPrivateKeyData(), KEY_ALGORITHM);
+
+            return new KeyPair(publicKey, privateKey);
+        }
+        catch (GeneralSecurityException e) {
+            // If any exception occurred, the keys are either malformed or not PKCS8 keys
+            log.debug("Unexpected exception occurred while parsing key data: ", e);
+            return null;
+        }
+    }
+
+    /**
+     * Attempts to process the given keypair data as if the keys are Java-serialized key objects. If
+     * the key pair data is incomplete or invalid, this method returns null.
+     *
+     * @param kpdata
+     *  the key pair data to process
+     *
+     * @return
+     *  a KeyPair consisting of the keys from the provided key pair data, or null if the key pair
+     *  data could not be processed
+     */
+    private KeyPair processAsJSO(KeyPairData kpdata) {
+        try {
+            PublicKey publicKey = this.deserializeKey(kpdata.getPublicKeyData(), PublicKey.class);
+            PrivateKey privateKey = this.deserializeKey(kpdata.getPrivateKeyData(), PrivateKey.class);
+
+            return new KeyPair(publicKey, privateKey);
+        }
+        catch (ClassNotFoundException | IOException e) {
+            // If any exception occurred, the keys are either malformed, not Java-serialized key
+            // objects, or something that doesn't allow extracting key data
+            log.debug("Unexpected exception occurred while deserializing key data: ", e);
+            return null;
+        }
+    }
+
+    /**
+     * Deserializes the given byte array into an object of the specified class. If the data provided
+     * cannot be deserialized into the given class, this method throws an exception.
+     *
+     * @param bytes
+     *  the data to deserialize
+     *
+     * @param keyClass
+     *  the key class to which the data should be deserialized
+     *
+     * @throws IOException
+     *  if the key data is malformed, invalid, or otherwise cannot be deserialized to the specified
+     *  class
```

Comment:
The javadoc should be updated to correspond with the code (IOException for null data, KeyPairCreationException for the rest)

Reply:
Done.